### PR TITLE
Enable scroll-view rendering optimization for array of children #5012

### LINKED ISSUE #5012: Crashes and CPU overload while fetching long history (>500 messages)
### User Story

As a user, I want to see all messages fetched correctly and I could use app (switch to profile, add new contacts etc.) simultaneously.

### Description
*Type*: Bug

*Summary*: reproducible with last `status-go develop` branch (#5001). When you join public channels with 7-day history > 500 messages (i.e., #status) then:
1) application hangs or crashes
2) StatusIm CPU Usage ~99-102 % 

IMPORTANT NOTE: can be reproducible with chats with 50-100 messages when you switching between chats during fetching messages or when you add new chat and switch to it when messages are not fetched yet. Not 100 % but I faced with it already 10+ times.

#### Expected behavior
app is responsive

#### Actual behavior
application hangs, then crashes
hhh

### Reproduction

- Open Status
- Join channel with history > 500 messages in 7 days (i.e.`#status`)
- Try to scroll message history or switch to profile; pay attention to CPU usage.

### Additional Information
[comment]: # (Please do your best to fill this out.)
* Status version: [PR-5001](https://jenkins.status.im/job/status-react/job/desktop/job/pull-requests/view/change-requests/job/PR-5001/)
* Operating System: MacOS High Sierra

#### Logs

* NOTE:  after crash you should manually exit `ubuntu-server` process an restart app to continue.
* Full StatusIm log: 
[StatusIm.log](https://github.com/status-im/status-react/files/2158713/StatusIm.log)

```
Process:               StatusIm [13208]
Path:                  /Users/*/Downloads/*/StatusIm.app/Contents/MacOS/StatusIm
Identifier:            StatusIm
Version:               0
Code Type:             X86-64 (Native)
Parent Process:        ??? [1]
Responsible:           StatusIm [13208]
User ID:               502

Date/Time:             2018-07-03 12:47:20.048 +0300
OS Version:            Mac OS X 10.13.5 (17F77)
Report Version:        12
Anonymous UUID:        93B7A170-78AE-1670-E0DF-EF45503C77C4

Sleep/Wake UUID:       5C4FB9E8-003B-4AC9-BA2D-C421DA3DB707

Time Awake Since Boot: 110000 seconds
Time Since Wake:       21000 seconds

System Integrity Protection: enabled

Crashed Thread:        0  CrBrowserMain  Dispatch queue: com.apple.main-thread

Exception Type:        EXC_BAD_ACCESS (SIGSEGV)
Exception Codes:       KERN_INVALID_ADDRESS at 0x00007fcf953ffff8
Exception Note:        EXC_CORPSE_NOTIFY

Termination Signal:    Segmentation fault: 11
Termination Reason:    Namespace SIGNAL, Code 0xb
Terminating Process:   exc handler [0]

VM Regions Near 0x7fcf953ffff8:
    MALLOC_SMALL           00007fcf93800000-00007fcf95000000 [ 24.0M] rw-/rwx SM=PRV  
--> 
    MALLOC_TINY            00007fcf95400000-00007fcf95800000 [ 4096K] rw-/rwx SM=PRV  

Thread 0 Crashed:: CrBrowserMain  Dispatch queue: com.apple.main-thread
0   libsystem_platform.dylib      	0x00007fff6ac7e169 _platform_memmove$VARIANT$Haswell + 585
1   StatusIm                      	0x0000000108c228ce 0x108bbf000 + 407758
2   StatusIm                      	0x0000000108c225a9 YGNode::insertChild(YGNode*, unsigned int) + 41
3   StatusIm                      	0x0000000108c14c5d YGNodeInsertChild + 125
....

```
 [full log](https://github.com/status-im/status-react/files/2158707/AppleCrashLog.log)

### PR DESCRIPTION
**Pull request by maxhora**
fixes https://github.com/status-im/status-react/issues/5012

### Summary:

Enables rendering optimization for scroll-view content

### Steps to test:
- Open Status
- Navigate to chat with longest history you have
- Scroll messages up; Scroll messages down

status: ready

### CODE REVIEW

File: src/status_im/ui/screens/desktop/main/chat/views.cljs [hunk 1]
```diff
@@ -135,16 +135,19 @@
...
       [react/view {:style styles/messages-view}
        [react/scroll-view {:scrollEventThrottle    16
+                           :headerHeight styles/messages-list-vertical-padding
+                           :footerWidth styles/messages-list-vertical-padding
+                           :enableArrayScrollingOptimization true
                            :on-scroll              (fn [e]
                                                      (let [ne (.-nativeEvent e)
                                                            y (.-y (.-contentOffset ne))]
-                                                       (when (zero? y)
+                                                       (when (<= y 0)
```
Comment (maxhora) [inline src/status_im/ui/screens/desktop/main/chat/views.cljs:144]: Negative y values are possible when view flicked down from the top most position. In terms of current implementation it means that user wants to load more messages from the history.

File: src/status_im/ui/screens/desktop/main/chat/views.cljs [hunk 2]
```diff
@@ -135,16 +135,19 @@
           current-public-key (re-frame/subscribe [:get-current-public-key])]
       [react/view {:style styles/messages-view}
        [react/scroll-view {:scrollEventThrottle    16
+                           :headerHeight styles/messages-list-vertical-padding
+                           :footerWidth styles/messages-list-vertical-padding
+                           :enableArrayScrollingOptimization true
```
Comment (maxhora) [inline src/status_im/ui/screens/desktop/main/chat/views.cljs:140]: `:enableArrayScrollingOptimization` option is introduced into react-native-desktop port to enable rendering optimization for lists of items inside ScrollView (items dynamically mounted and unmounted from the scene graph)

**Review by vkjr — APPROVED** [review]

**Review by yenda — APPROVED** [review]

**Comment by maxhora** [comment]
@churik link to Desktop binaries for verification https://ci.status.im/job/status-react/job/desktop/job/pull-requests/job/PR-5431/3/

<blockquote><div><strong><a href="https://ci.status.im/job/status-react/job/desktop/job/pull-requests/job/PR-5431/3/">status-react » desktop » pull-requests » PR-5431 #3 [Jenkins]</a></strong></div></blockquote>

**Comment by churik** [comment]
Tested (IOS, Android):
- creating account
- joining to #status  chat;
- scrolling history;


Tested (Mac OSx):
- creating account
- joining to several chats with 2+ day history;
- scrolling history;

**Comment by maxhora** [comment]
Final verification of merged changes was done on branch `optimization/scroll-view-rebased` rebased against branch `dmitryn/fix-desktop-deps` to pass Desktop build failures. Build artifacts available under https://ci.status.im/job/status-react/job/desktop/job/manual/124/

<blockquote><div><strong><a href="https://ci.status.im/job/status-react/job/desktop/job/manual/124/">status-react » desktop » manual #124 [Jenkins]</a></strong></div></blockquote>

**Pull request merged**